app/scripts/lib/snap-keyring/snap-keyring.test.ts

```diff
@@ -1,166 +1,412 @@
+        method: 'notify:accountCreated',
+        params: {
+          account: mockAccount,
+          displayConfirmation: true,
+          accountNameSuggestion: mockNameSuggestion,
+        },
+      });
 
-      expect(controllerMessenger.call).toHaveBeenNthCalledWith(
-        1,
-        'ApprovalController:startFlow',
-      );
-      expect(controllerMessenger.call).toHaveBeenNthCalledWith(
-        2,
-        'ApprovalController:addRequest',
+      expect(mockStartFlow).toHaveBeenCalledTimes(1);
+      // First request for show account creation dialog
+      // Second request for account name suggestion second
+      expect(mockAddRequest).toHaveBeenCalledTimes(2);
+      expect(mockAddRequest).toHaveBeenNthCalledWith(1, [
+        {
+          origin: mockSnapId,
+          type: SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.confirmAccountCreation,
+        },
+        true,
+      ]);
+      // First call is from addAccount after user confirmation
+      // Second call is from within the SnapKeyring
+      expect(mockPersisKeyringHelper).toHaveBeenCalledTimes(2);
+      expect(mockAddRequest).toHaveBeenNthCalledWith(2, [
         {
           origin: mockSnapId,
           type: SNAP_MANAGE_ACCOUNTS_CONFIRMATION_TYPES.showNameSnapAccount,
           requestData: {
-            address,
-            snapSuggestedAccountName: accountNameSuggestion,
+            address: mockInternalAccount.address.toLowerCase(),
+            snapSuggestedAccountName: mockNameSuggestion,
           },
         },
         true,
+      ]);
+      expect(mockGetAccountByAddress).toHaveBeenCalledTimes(1);
+      expect(mockGetAccountByAddress).toHaveBeenCalledWith([
+        mockAccount.address.toLowerCase(),
+      ]);
+      expect(mockTrackEvent).toHaveBeenCalledTimes(3);
+      expect(mockTrackEvent).toHaveBeenNthCalledWith(1, {
+        category: MetaMetricsEventCategory.Accounts,
+        event: MetaMetricsEventName.AddSnapAccountSuccessViewed,
+        properties: {
+          account_type: 'snap',
+          snap_id: mockSnapId,
+          snap_name: mockSnapName,
+        },
+      });
+      expect(mockTrackEvent).toHaveBeenNthCalledWith(2, {
+        category: MetaMetricsEventCategory.Accounts,
+        event: MetaMetricsEventName.AddSnapAccountSuccessClicked,
+        properties: {
+          account_type: 'snap',
+          snap_id: mockSnapId,
+          snap_name: mockSnapName,
+        },
+      });
+      expect(mockTrackEvent).toHaveBeenNthCalledWith(3, {
+        category: MetaMetricsEventCategory.Accounts,
+        event: MetaMetricsEventName.AccountAdded,
+        properties: {
+          account_type: 'snap',
+          snap_id: mockSnapId,
+          snap_name: mockSnapName,
+        },
+      });
+      expect(mockShowSuccess).toHaveBeenCalledTimes(1);
+      expect(mockEndFlow).toHaveBeenCalledTimes(1);
+      expect(mockEndFlow).toHaveBeenCalledWith([{ id: mockFlowId }]);
+    });
```

Comment:
Those new tests are really good, however, I feel like we duplicate a lot of codes in there...

I wonder if we should use a `it.each` here with different `params` object for the `handleKeyringSnapMessage` and use those parameters to conditionally checks for some part of the flow (like `displayConfirmation: true`, with/without`accountNameSuggestion`)

This way it would be easier to use the same "flow structure" for this (this would also reduce the test code too).

WDYT?

Reply:
Yea there alot of duplicate code but its because doesn't recommend conditional assertions